packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
-        t = new Enum(type);
+        return new Enum(parent, type);
       }
+      return unknownType(type);
+    },
+  );
+}
 
-      if (t) {
-        this.types[type.fqn] = t;
-      }
-    });
+function findTypeInTree(module: Module, fqn: string): ModuleType | undefined {
+  const result = module.types.find((t) => t.type.fqn === fqn);
 
-    this.submodules = this.buildSubmodules(assembly, submodule);
+  if (result) {
+    return result;
+  } else if (module.hasSubmodules()) {
+    return module.submodules.reduce((accum: ModuleType | undefined, sm) => {
+      return accum || findTypeInTree(sm, fqn);
+    }, undefined);
   }
 
-  public buildSubmodules(
-    assembly: Assembly,
-    submodule?: JsiiSubmodule,
-  ): Submodule[] {
-    return (submodule?.submodules ?? assembly.submodules).map(
-      (sm) => new Submodule(assembly, sm),
+  return undefined;
+}
+
+const JSII_MODULE_NAME = 'github.com/aws-cdk/jsii/jsii';
+export abstract class ModuleFile {
+  public constructor(public readonly file: string) {}
+  public open(code: CodeMaker): void {
+    code.openFile(this.file);
+  }
+
+  public close(code: CodeMaker): void {
+    code.closeFile(this.file);
+  }
+}
+
+export class RootModule extends ModuleFile implements Module {
+  private readonly assembly: Assembly;
+  public readonly packageName: string;
+  public readonly moduleName: string;
+  public readonly file: string;
+  public readonly filePath: string;
+  public readonly root: RootModule;
+
+  public constructor(assembly: Assembly) {
+    const packageName = assembly.name
+      .replace('@', '')
+      .replace(/[^a-z0-9_.]/gi, '');
+    const filePath = join(...packageName.split('.'));
+    const file = `${filePath}.go`;
+    super(file);
+
+    this.assembly = assembly;
+    this.root = this;
+    this.packageName = packageName;
+    // moduleName == packageName for root;
+    this.moduleName = packageName;
+    this.filePath = filePath;
+    this.file = file;
+  }
+
+  public get types(): ModuleTypes {
+    return buildModuleTypes(this, Object.values(this.assembly.types));
+  }
+
+  public get submodules(): Submodule[] {
+    return buildSubmodules(this, this, this.assembly.submodules);
+  }
+
+  public get dependencies(): Set<Module> {
+    return new Set(
+      this.types
+        .reduce(
+          (accum: Module[], t: ModuleType) => [...accum, ...t.dependencies],
+          [],
+        )
+        .filter((mod) => mod.packageName !== this.packageName),
     );
   }
 
+  public hasSubmodules(): boolean {
+    return Boolean(this.submodules.length);
+  }
+
+  public findType(fqn: string): ModuleType | undefined {
+    return findTypeInTree(this, fqn);
+  }
+
+  public emit(code: CodeMaker): void {
+    this.open(code);
+    code.line(`package ${this.packageName}`);
+    code.line();
+
+    code.line('import (');
+    code.indent();
+    code.line(`"${JSII_MODULE_NAME}"`);
+    if (this.dependencies.size > 0) {
+      this.dependencies.forEach((mod) => {
```

Comment:
Why nest pithing this `if` since the empty collection's `forEach` will immediately return?

Also - `for (const mod of this.dependencies)` is preferred.

Reply:
I can't remember why but I definitely had a reason at some point... 🤷‍♂️